src/Build/Graph/GraphBuilder.cs

```diff
@@ -110,28 +111,74 @@ private static IReadOnlyCollection<ProjectGraphNode> GetGraphRoots(IReadOnlyColl
+
+            _projectInterpretation.ReparentInnerBuilds(allParsedProjects, this);
+
             if (_solutionDependencies != null && _solutionDependencies.Count != 0)
             {
                 AddEdgesFromSolution(allParsedProjects, _solutionDependencies, Edges);
             }
         }
 
-        private static void AddEdgesFromProjectFileReferences(Dictionary<ConfigurationMetadata, ParsedProject> allParsedProjects, GraphEdges edges)
+        private void AddEdgesFromProjectReferenceItems(Dictionary<ConfigurationMetadata, ParsedProject> allParsedProjects, GraphEdges edges)
         {
+            var transitiveReferenceCache = new Dictionary<ProjectGraphNode, IEnumerable<ProjectGraphNode>>(allParsedProjects.Count);
+
             foreach (var parsedProject in allParsedProjects)
             {
+                var currentNode = parsedProject.Value.GraphNode;
+
+                var requiresTransitiveProjectReferences = _projectInterpretation.RequiresTransitiveProjectReferences(currentNode.ProjectInstance);
+
                 foreach (var referenceInfo in parsedProject.Value.ReferenceInfos)
                 {
-                    ErrorUtilities.VerifyThrow(
-                        allParsedProjects.ContainsKey(referenceInfo.ReferenceConfiguration),
-                        "all references should have been parsed");
-
-                    parsedProject.Value.GraphNode.AddProjectReference(
+                    // Always add direct references.
+                    currentNode.AddProjectReference(
                         allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode,
                         referenceInfo.ProjectReferenceItem,
                         edges);
+
+                    // Add transitive references only if the project requires it.
+                    if (requiresTransitiveProjectReferences)
+                    {
+                        foreach (var transitiveProjectReference in GetTransitiveProjectReferences(allParsedProjects[referenceInfo.ReferenceConfiguration]))
+                        {
+                            currentNode.AddProjectReference(
+                                transitiveProjectReference,
+                                new ProjectItemInstance(
+                                    project: currentNode.ProjectInstance,
+                                    itemType: ProjectInterpretation.TransitiveReferenceItemName,
+                                    includeEscaped: referenceInfo.ReferenceConfiguration.ProjectFullPath,
+                                    directMetadata: null,
+                                    definingFileEscaped: currentNode.ProjectInstance.FullPath
+                                ),
+                                edges);
+                        }
+                    }
+                }
+            }
+
+            IEnumerable<ProjectGraphNode> GetTransitiveProjectReferences(ParsedProject parsedProject)
```

Comment:
```suggestion
            List<ProjectGraphNode> GetTransitiveProjectReferences(ParsedProject parsedProject)
```

So that traversing it multiple times doesn't get slow?